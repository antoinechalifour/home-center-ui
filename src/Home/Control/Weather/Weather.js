import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { graphql } from 'react-apollo'
import gql from 'graphql-tag'
import Loader from 'ui/Loader'
import WindSpeedIcon from 'react-icons/lib/ti/weather-windy'
import ThermometerIcon from 'react-icons/lib/ti/thermometer'
import ClearIcon from 'react-icons/lib/ti/weather-partly-sunny'
import CloudsIcon from 'react-icons/lib/ti/weather-cloudy'
import MistIcon from 'react-icons/lib/ti/waves'

const Wrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 24px;
`

const WeatherIconWrapper = styled.div`
  margin-right: 24px;

  svg {
    font-size: 128px;
  }
`

const Temp = styled.div`
  font-size: 64px;
`

const TempMin = styled.span`
  opacity: .45;
`

const Info = styled.div`
  span + span::before {
    content: '•';
    margin: auto 12px;
    opacity: .54;
  }

  svg {
    font-size: 18px;
  }
`

const KindToIcon = {
  clear: ClearIcon,
  clouds: CloudsIcon,
  mist: MistIcon,
  fog: MistIcon
}

const Weather = ({ latitude, longitude, data }) => {
  if (data.loading) {
    return <Loader />
  }

  const weather = data.weather
  console.log(weather.kind)
  const WeatherIcon = KindToIcon[weather.kind]

  return (
    <Wrapper>
      <WeatherIconWrapper>
        <WeatherIcon />
      </WeatherIconWrapper>
      <div>
        <Temp>{weather.temp}°</Temp>
        <Info>
          <span>
            {weather.city}
          </span>
          <span>
            <WindSpeedIcon /> {weather.wind_speed} km/h
          </span>
          <span>
            <ThermometerIcon />
            {' '}
            <TempMin>{weather.temp_min}</TempMin>
            {' '}
            -
            {' '}
            {weather.temp_max}
          </span>
        </Info>
      </div>
    </Wrapper>
  )
}

Weather.propTypes = {
  latitude: PropTypes.number.isRequired,
  longitude: PropTypes.number.isRequired,
  data: PropTypes.shape({
    loading: PropTypes.bool.isRequired,
    weather: PropTypes.shape({
      city: PropTypes.string.isRequired,
      kind: PropTypes.string.isRequired,
      pressure: PropTypes.number.isRequired,
      temp: PropTypes.number.isRequired,
      temp_min: PropTypes.number.isRequired,
      temp_max: PropTypes.number.isRequired,
      wind_speed: PropTypes.number.isRequired
    })
  }).isRequired
}

export default graphql(
  gql`
  query WeatherQuerr ($lon: Float, $lat: Float) {
    weather (lon: $lon, lat: $lat) {
      city,
      kind
      temp
      temp_min
      temp_max
      pressure
      wind_speed
    }
  }
`,
  {
    options: ({ longitude, latitude }) => ({
      variables: { lon: longitude, lat: latitude }
    })
  }
)(Weather)
