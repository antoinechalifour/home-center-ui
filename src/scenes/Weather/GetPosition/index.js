import { Component } from 'react'
import PropTypes from 'prop-types'

export default class GetPosition extends Component {
  static propTypes = {
    render: PropTypes.func.isRequired
  }

  state = { position: null, err: null }

  onPosition = position => {
    const { latitude, longitude } = position.coords

    this.setState({ position: { latitude, longitude } })
  }

  onError = err => {
    this.setState({ err })
  }

  componentDidMount () {
    navigator.geolocation.getCurrentPosition(this.onPosition, this.onError)
  }

  render () {
    return this.props.render({
      err: this.state.err,
      position: this.state.position
    })
  }
}
