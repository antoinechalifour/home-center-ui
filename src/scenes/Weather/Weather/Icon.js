import React from 'react'
import PropTypes from 'prop-types'
import ClearIcon from 'react-icons/lib/ti/weather-partly-sunny'
import CloudsIcon from 'react-icons/lib/ti/weather-cloudy'
import MistIcon from 'react-icons/lib/ti/waves'
import RainIcon from 'react-icons/lib/ti/weather-shower'
import SnowIcon from 'react-icons/lib/ti/weather-snow'

const typeToIcon = {
  clear: ClearIcon,
  clouds: CloudsIcon,
  mist: MistIcon,
  fog: MistIcon,
  rain: RainIcon,
  snow: SnowIcon
}

export default function Icon ({ type }) {
  const WeatherIcon = typeToIcon[type]

  if (!WeatherIcon) {
    console.warn(`Missing icon for weather type "${type}"`)
    return null
  }

  return <WeatherIcon />
}

Icon.propTypes = {
  type: PropTypes.string.isRequired
}
