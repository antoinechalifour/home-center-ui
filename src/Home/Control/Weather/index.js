import React, { Component } from 'react'
import Weather from './Weather'

class GetPosition extends Component {
  state = { position: null }

  componentDidMount () {
    navigator.geolocation.getCurrentPosition(this.onPosition, this.onError)
  }

  onPosition = data => {
    const coords = data.coords
    const { latitude, longitude } = coords

    this.setState({
      position: { latitude, longitude }
    })
  }

  onError = err => {
    console.log(err)
  }

  render () {
    return this.state.position ? <Weather {...this.state.position} /> : null
  }
}

export default GetPosition
