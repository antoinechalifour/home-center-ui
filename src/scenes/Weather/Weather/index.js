import React, { Component } from 'react'
import { graphql } from 'react-apollo'
import PropTypes from 'prop-types'
import { getWeather } from 'queries/weather'
import Weather from './Weather'

class WeatherContainer extends Component {
  static propTypes = {
    latitude: PropTypes.number,
    longitude: PropTypes.number,
    data: PropTypes.shape({
      startPolling: PropTypes.func.isRequired,
      stopPolling: PropTypes.func.isRequired
    }).isRequired
  }

  componentDidMount () {
    this.props.data.startPolling(60000)
  }

  componentWillUnmount () {
    this.props.data.stopPolling()
  }

  render () {
    return <Weather {...this.props} />
  }
}

const queryOptions = {
  options: ({ longitude, latitude }) => ({
    variables: { lon: longitude, lat: latitude }
  })
}

export default graphql(getWeather, queryOptions)(WeatherContainer)
