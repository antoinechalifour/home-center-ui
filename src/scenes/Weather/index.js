import React from 'react'
import Loader from 'ui/Loader'
import GetPosition from 'components/GetPosition'
import Weather from './Weather'

export default function WeatherWidget () {
  return (
    <GetPosition
      render={({ err, position }) => {
        const isLoading = !err && !position

        if (isLoading) {
          return <Loader />
        } else if (err) {
          return <div>Error !</div>
        } else {
          return <Weather {...position} />
        }
      }}
    />
  )
}
