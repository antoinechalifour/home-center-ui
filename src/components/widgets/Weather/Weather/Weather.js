import React, { Component, Fragment } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import WindSpeedIcon from 'react-icons/lib/ti/weather-windy'
import ThermometerIcon from 'react-icons/lib/ti/thermometer'
import { getDayName } from 'util/dates'
import { getWeatherName } from 'util/weather'
import gqlLoaderHoc from 'components/GqlLoader'
import Icon from './Icon'
import Forecast from './Forecast'

class Clock extends Component {
  state = {
    date: new Date()
  }

  componentDidMount () {
    this._interval = window.setInterval(this._updateClock, 1000)
  }

  componentWillUnmount () {
    window.clearInterval(this._interval)
  }

  _updateClock = () => {
    this.setState({ date: new Date() })
  }

  render () {
    let hours = this.state.date.getHours()
    let minutes = this.state.date.getMinutes()

    hours = hours < 10 ? `0${hours}` : hours
    minutes = minutes < 10 ? `0${minutes}` : minutes

    return <div>{hours}:{minutes}</div>
  }
}

export class Weather extends Component {
  _renderTop () {
    const weather = this.props.data.weather

    return (
      <Fragment>
        <Icon type={weather.kind} />
        <Title>Current weather</Title>
        <WeatherType>{getWeatherName(weather.kind)}</WeatherType>
        <Geolocation>
          <div>{weather.city}</div>
          <Clock />
        </Geolocation>
      </Fragment>
    )
  }

  _renderDetail () {
    return (
      <Detail>
        <Temperature>
          <div>
            {Math.round(this.props.data.weather.temp)}<TempSign>Â°</TempSign>
          </div>
          <div>{getDayName(new Date())}</div>
        </Temperature>
        <WindInformation>
          <div>
            <WindSpeedIcon />
          </div>
          <div>{this.props.data.weather.wind_speed}km/h</div>
        </WindInformation>
      </Detail>
    )
  }

  render () {
    return (
      <Container>
        <Top>
          {this._renderTop()}
        </Top>
        <Bottom>
          {this._renderDetail()}
          <Forecast />
        </Bottom>
      </Container>
    )
  }
}

Weather.propTypes = {
  data: PropTypes.shape({
    weather: PropTypes.shape({
      city: PropTypes.string.isRequired,
      kind: PropTypes.string.isRequired,
      pressure: PropTypes.number.isRequired,
      temp: PropTypes.number.isRequired,
      temp_min: PropTypes.number.isRequired,
      temp_max: PropTypes.number.isRequired,
      wind_speed: PropTypes.number.isRequired
    })
  }).isRequired
}

export default gqlLoaderHoc(Weather)

const Container = styled.div`
  clear: both;
`

const Top = styled.div`
  padding: 32px 12px;
  position: relative;
  text-align: center;
  background: ${({ theme }) => theme.colors.primary};
  color: #fff;

  svg {
    font-size: 80px;
  }
`

const Title = styled.div`
  position: absolute;
  top: 12px;
  left: 12px;
`

const WeatherType = styled.div`
  position: absolute;
  bottom: 12px;
  left: 12px;
`

const Geolocation = styled.div`
  position: absolute;
  right: 12px;
  top: 12px;
  text-align: right;
`

const Bottom = styled.div`
  width: 100%;
  float: left;
  display: flex;
  overflow-x: auto;
`

const Detail = styled.div`
  display: flex;
  align-items: center;
`

const Temperature = styled.div`
  flex: 1 0 55px;
  padding: 12px;
  text-align: center;

  > div:first-child {
    font-size: 32px;
  }

  > div:last-child {
    opacity: .54;
  }
`

const WindInformation = styled.div`
  flex: 1 0 55px;
  padding: 12px;
  text-align: center;
  position: relative;
  top: 6px;

  svg {
    font-size: 32px;
    margin-bottom: 8px;
  }

  > div:last-child {
    opacity: .54;
    font-size: 12px;
  }
`

// const ForecastContainer = styled.div`
//   flex: 1;
//   border-left: 1px solid rgba(0, 0, 0, .15);
//   overflow-x: auto;

//   li {
//     flex: 1 0 55px;
//   }
// `

const TempSign = styled.span`
  font-size: 65%;
  opacity: .54;
  vertical-align: top;
  position: relative;
  top: 4px;
  left: 2px;
`
