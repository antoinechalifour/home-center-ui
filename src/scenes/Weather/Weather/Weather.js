import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import WindSpeedIcon from 'react-icons/lib/ti/weather-windy'
import ThermometerIcon from 'react-icons/lib/ti/thermometer'
import Loader from 'ui/Loader'
import Col from 'ui/Col'
import Row from 'ui/Row'
import Card, * as card from 'ui/Card'
import Icon from './Icon'

export default function Weather ({ data }) {
  if (data.loading) {
    return <Loader />
  }

  const weather = data.weather

  return (
    <Card>
      <card.Content>
        <Col align='center'>
          <MainRow align='center'>
            <Icon type={weather.kind} />
            <div>{weather.temp}°</div>
          </MainRow>
          <DetailRow align='center'>
            <span>
              {weather.city}
            </span>
            <span>
              <WindSpeedIcon /> {weather.wind_speed} km/h
            </span>
            <span>
              <ThermometerIcon />
              {' '}
              <MinTemperature>{weather.temp_min}</MinTemperature>
              {' '}
              -
              {' '}
              {weather.temp_max}
            </span>
          </DetailRow>
        </Col>
      </card.Content>
    </Card>
  )
}

Weather.propTypes = {
  data: PropTypes.shape({
    loading: PropTypes.bool.isRequired,
    weather: PropTypes.shape({
      city: PropTypes.string.isRequired,
      kind: PropTypes.string.isRequired,
      pressure: PropTypes.number.isRequired,
      temp: PropTypes.number.isRequired,
      temp_min: PropTypes.number.isRequired,
      temp_max: PropTypes.number.isRequired,
      wind_speed: PropTypes.number.isRequired
    })
  }).isRequired
}

const MainRow = styled(Row)`
  font-size: 64px;

  > div:last-child {
    margin-left: 24px;
  }
`

const DetailRow = styled(Row)`
  opacity: .54;
  
  span + span::before {
    content: '•';
    margin-left: 12px;
    margin-right: 12px;
    opacity: .5;
  }
`

const MinTemperature = styled.span`
  opacity: .54;
`
