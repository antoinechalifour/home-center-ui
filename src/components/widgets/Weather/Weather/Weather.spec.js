import React from 'react'
import { shallow } from 'enzyme'
import { Weather } from './Weather'

describe('scenes/Weather/Weather', () => {
  it('Should render the widget', () => {
    const props = {
      data: {
        loading: false,
        weather: {
          city: 'Limoges',
          kind: 'rain',
          pressure: 1008,
          temp: 6,
          temp_max: 6,
          temp_min: 6,
          wind_speed: 7.2
        }
      }
    }
    const wrapper = shallow(<Weather {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})
