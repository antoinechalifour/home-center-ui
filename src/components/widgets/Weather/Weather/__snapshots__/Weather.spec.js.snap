// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scenes/Weather/Weather Should render the widget 1`] = `
<div>
  <styled.div>
    <Icon
      type="rain"
    />
    <div>
      6
      Â°
    </div>
  </styled.div>
  <styled.div>
    <span>
      Limoges
    </span>
    <span>
      <TiWeatherWindy />
       
      7.2
       km/h
    </span>
    <span>
      <TiThermometer />
       
      <styled.span>
        6
      </styled.span>
       
      -
       
      6
    </span>
  </styled.div>
</div>
`;
